### PR TITLE
Add basic authentication in the server

### PR DESCRIPTION
**Pull request by bmccandless**
A pattern for creating authentication methods is introduced, with three
authentication types defined:
  none - no authentication
  session - like the current session based auth used for user annotations
  test - used to test the login/logout process end to end

The config endpoint now returns informations about the authentication, like if
the user is authenticated and their username.  The redirect uri's for login and
logout are also returned if the authentication type requires login

This is the first a several PRs for authentication.

### CODE REVIEW

File: server/app/app.py [hunk 1]
```diff
@@ -160,6 +162,17 @@ def dataroot_test_index():
 
     config = current_app.app_config
     server_config = config.server_config
+
+    auth = server_config.auth
+    if auth.is_valid():
+        if server_config.auth.is_authenticated():
```
Comment (MDunitz) [inline server/app/app.py:168]:
```suggestion
        if auth.is_authenticated():
```

File: server/app/app.py [hunk 2]
```diff
@@ -160,6 +162,17 @@ def dataroot_test_index():
...
+            data += f"<p>Logged in as {auth.get_userid()} / {auth.get_username()}</p>"
+        if auth.requires_client_login():
+            if server_config.auth.is_authenticated():
```
Comment (MDunitz) [inline server/app/app.py:171]:
```suggestion
            if auth.is_authenticated():
```

File: server/common/app_config.py [hunk 3]
```diff
@@ -414,8 +432,12 @@ def __init__(self, app_config, default_config):
         # The matrix data cache manager is created during the complete_config and stored here.
         self.matrix_data_cache_manager = None
 
+        # The authentication object (BCM -- better name)
```
Comment (MDunitz) [inline server/common/app_config.py:435]: could you write this as ```# TODO @brian better name```

Comment (bmccandless) [reply server/common/app_config.py:435]: Oops, didn't mean to commit the "BCM".   I'll fix this and the other issue in my next PR.

**Pull request merged**